### PR TITLE
Send attendence links to required and optinal attendees of an event without an RSVP

### LINKED ISSUE #21189: Calendar invitation accept/decline buttons missing
I just received a calender invitation and it seems like the accept/decline buttons are missing in the latest version (Nextcloud 19, Calendar 2.0.3):

image

### PR DESCRIPTION
**Pull request by miaulalala**
To compensate for the missing RSVP from Mail Clients like Apple Mail and Thunderbird, also look at the role of the attendee to determine if they should receive a link for the event.

Fixes #21189

### CODE REVIEW

**Comment by miaulalala** [comment]
@ChristophWurst I'm not sure if adding the OPT-PARTICIPANT ROLE to the check would be beneficial. What do you think?

File: apps/dav/lib/CalDAV/Appointments/Plugin.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OCA\DAV\CalDAV\Appointments;
+
+/**
+ * Default built-in types provided by Doctrine DBAL.
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/Appointments/Plugin.php:8]: :see_no_evil:

File: apps/dav/lib/CalDAV/Appointments/Plugin.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	public const SAT = 'SAT';
+	public const SUN = 'SUN';
+	public const EVERY_X = '%s %s %s';
+
+	/**
+	 * @codeCoverageIgnore
+	 */
+	private function __construct()
+	{
+	}
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/Appointments/Plugin.php:29]:
```suggestion
```

File: apps/dav/lib/CalDAV/Schedule/IMipPlugin.php [hunk 3]
```diff
@@ -395,12 +394,18 @@ private function getAttendeeLangOrDefault($default, Property $attendee = null) {
...
+	private function getAttendeeRsvpOrReqParticipant(Property $attendee = null) {
 		if ($attendee !== null) {
 			$rsvp = $attendee->offsetGet('RSVP');
 			if (($rsvp instanceof Parameter) && (strcasecmp($rsvp->getValue(), 'TRUE') === 0)) {
 				return true;
 			}
+			$role = $attendee->offsetGet('ROLE');
+			// @see https://datatracker.ietf.org/doc/html/rfc5545#section-3.2.16
+			// Attendees without a role are assumed required and should receive an invitation link even if they have no RSVP set
+			if($role === null || (($rsvp instanceof Parameter) && (strcasecmp($role->getValue(), 'REQ-PARTICIPANT') === 0))) {
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:406]:
```suggestion
			if($role === null || (($role instanceof Parameter) && (strcasecmp($role->getValue(), 'REQ-PARTICIPANT') === 0))) {
```

**Review by ChristophWurst — APPROVED** [review]
But remove the unrelated new file :)

**Comment by ChristophWurst** [comment]
> @ChristophWurst I'm not sure if adding the OPT-PARTICIPANT ROLE to the check would be beneficial. What do you think?

so that optional participants also get an invite when no RSVP is set?

**Comment by miaulalala** [comment]
@ChristophWurst yes - I'm not sure if this is neccessary or even wanted, but it makes sense to me.

**Comment by ChristophWurst** [comment]
some insights on the seemingly random psalm security check failure

``Psalm dev-master@eb973ab2e1877845c74d65a9c6064fda9e34375c`` passes
``Psalm dev-master@a655ca81c15e3b3fdc9c8c55a04d9f036c897b16`` fails

**Review by ChristophWurst — APPROVED** [review]
Tested & works

File: apps/dav/lib/CalDAV/Schedule/IMipPlugin.php [hunk 4]
```diff
@@ -395,12 +394,18 @@ private function getAttendeeLangOrDefault($default, Property $attendee = null) {
 	 * @param Property|null $attendee
 	 * @return bool
 	 */
-	private function getAttendeeRSVP(Property $attendee = null) {
```
[the remark this answers is not in the document]
Comment (Shar1988) [reply apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:398]: nll

**Review by Shar1988 — APPROVED** [review]

**Review by Shar1988 — APPROVED** [review]

**Comment by Shar1988** [comment]
> To compensate for the missing RSVP from Mail Clients like Apple Mail and Thunderbird, also look at the role of the attendee to determine if they should receive a link for the event.
> 
> Fixes #21189

56cc01c45c5d34938c9affa190df9e479690bad1

**Review by kesselb — APPROVED** [review]

**Pull request merged**